Exposes _aminmax as aminmax and makes it structured

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62432
* __->__ #62401

This PR exposes the `torch._aminmax` operator as `torch.aminmax`.

**TODO**

- [x] add examples to documentation
- [x] add minmax to rst docs

fixes https://github.com/pytorch/pytorch/issues/62164

Differential Revision: [D30072246](https://our.internmc.facebook.com/intern/diff/D30072246)

heitorschueroff committed:
Exposes _aminmax as minmax

[ghstack-poisoned]

heitorschueroff committed:
fix lint on "Exposes _aminmax as minmax"


This PR exposes the `torch._aminmax` operator as `torch.minmax`.

**TODO**

- [ ] implement backwards

[ghstack-poisoned]

heitorschueroff commented:
@albanD How should I add the derivative for this function as it returns two output tensors?

ezyang commented:
```
#   - 'grad', the gradient of the output (often spelled grad_output
#     in Python) which we are going to left-multiply.
#
#     When a function returns multiple *differentiable* outputs,
#     you can refer to the gradients of each outputs using 'grads',
#     e.g., 'grads[0]', 'grads[1]'.
#
#     When a function returns *one* differentiable output (the
#     first output) and some more nondifferentiable outputs,
#     you MUST refer to the gradient of the differentiable output with
#     'grad' (this case is special-cased in our code generation).
#
#     Note that the number of differentibale outputs can
```

ngimel commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -118,6 +118,65 @@ TORCH_META_FUNC(argmin)
+    auto dim = maybe_wrap_dim(dim_opt.value(), self.ndimension());
+    native::zero_numel_check_dims(self, dim, "minmax");
+    shape = get_reduction_shape(self, dim, keepdim);
+  } else if (keepdim) {
```
what happens if `dim_opt` is empty and `keepdim` is false?

heitorschueroff replied:
The same wrong behavior that most reduction ops have now, it will reduce everything. Good catch! I'll update it to a noop.

ngimel replied:
But it shouldn't be a no-op. `x.minmax()` (which is empty dim_opt and `false` keepdim) should reduce all dimensions. It's just with your logic it never checks for `self.numel() > 0` and doesn't set 0-d tensor `shape` (I don't know if `set_output` can deal with uninitialized DimeVector argument).

heitorschueroff replied:
Oh sorry, I misunderstood you. Yes, you are right! empty dim_opt is when `dim=None` which should reduce all dimensions. `dim=[]` should be a noop. Thanks for catching this, I'll update it accordingly.

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2701,16 +2701,29 @@
   dispatch:
     CPU, CUDA: _aminmax
 
+- func: minmax(Tensor self, *, int? dim=None, bool keepdim=False) -> (Tensor min, Tensor max)
```
Why minmax and not aminmax?
Is there a numpy sister function that we match here?

cc @mruberry

ngimel replied:
No, numpy (and python array API) don't have minmax.

albanD replied:
And is this function behavior the same as the existing `min`/`max` or existing `amin`/`amax`?

heitorschueroff replied:
From offline discussion with @mruberry, I will update this to torch.aminmax.

heitorschueroff committed:
changed minmax to aminmax on "Exposes _aminmax as minmax"


This PR exposes the `torch._aminmax` operator as `torch.minmax`.

**TODO**

- [ ] implement backwards
- [ ] add examples to documentation
- [ ] add minmax to rst docs

fixes https://github.com/pytorch/pytorch/issues/62164

[ghstack-poisoned]

heitorschueroff committed:
add examples in docs on "Exposes _aminmax as aminmax"


This PR exposes the `torch._aminmax` operator as `torch.aminmax`.

**TODO**

- [ ] add examples to documentation
- [x] add minmax to rst docs

fixes https://github.com/pytorch/pytorch/issues/62164

[ghstack-poisoned]

heitorschueroff committed:
Update on "Exposes _aminmax as aminmax"


This PR exposes the `torch._aminmax` operator as `torch.aminmax`.

**TODO**

- [x] add examples to documentation
- [x] add minmax to rst docs

fixes https://github.com/pytorch/pytorch/issues/62164

[ghstack-poisoned]

heitorschueroff committed:
Update on "Exposes _aminmax as aminmax"


This PR exposes the `torch._aminmax` operator as `torch.aminmax`.

**TODO**

- [x] add examples to documentation
- [x] add minmax to rst docs

fixes https://github.com/pytorch/pytorch/issues/62164

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -118,6 +118,67 @@ TORCH_META_FUNC(argmin)
+  } else {
+    TORCH_CHECK(
+        self.numel() > 0,
+        "minmax(): cannot compute minmax over an empty dimension as the "
```
"minmax" -> "aminmax"

mruberry commented on aten/src/ATen/native/cpu/TensorCompareKernel.cpp:
```diff
@@ -161,20 +162,20 @@ static void max_kernel_impl(
     " but got ", min_result.scalar_type(), " and ", max_result.scalar_type());
 
-  AT_DISPATCH_ALL_TYPES_AND(ScalarType::Bool, self.scalar_type(), "_aminmax_cpu", [&] {
+  AT_DISPATCH_ALL_TYPES_AND(ScalarType::Bool, self.scalar_type(), "aminmax_cpu", [&] {
```
Fun fact: changing the name of a kernel might break mobile, but I don't think any mobile network uses this kernel so this should be OK

mruberry commented on aten/src/ATen/native/cuda/ReduceMinMaxKernel.cu:
```diff
@@ -117,15 +117,15 @@ static void max_kernel_impl(Tensor& result, Tensor& indice, const Tensor& self,
+  at::TensorIterator iter = make_reduction("minmax_cuda", min_result,
     max_result, self, dim, keepdim, self.scalar_type());
-  AT_DISPATCH_ALL_TYPES_AND2(kHalf, kBool, self.scalar_type(), "_aminmax_cuda", [&]() {
+  AT_DISPATCH_ALL_TYPES_AND2(kHalf, kBool, self.scalar_type(), "minmax_cuda", [&]() {
```
"minmax" -> "aminmax" here and on line 126 above

mruberry commented on aten/src/ATen/native/cuda/ReduceMinMaxKernel.cu:
```diff
@@ -117,15 +117,15 @@ static void max_kernel_impl(Tensor& result, Tensor& indice, const Tensor& self,
-static void _aminmax_kernel_impl(
-    Tensor& min_result,
-    Tensor& max_result,
+static void minmax_kernel_impl(
```
aminmax_kernel_impl

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1865,12 +1865,29 @@ def sample_inputs_amax_amin(op_info, device, dtype, requires_grad, **kwargs):
+                             args=args, kwargs=kwargs)
+                 for size, args, kwargs in test_cases)
+
+# TODO (@heitorschueroff) Once minmax supports multiple dims this should
```
"Once aminmax..."

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6374,6 +6391,16 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+           supports_autograd=False,
+           sample_inputs_func=sample_inputs_aminmax,
+           skips=(
+               # FIXME: minmax does not check for safe casting to output
```
"aminmax"

mruberry commented on aten/src/ATen/native/cuda/ReduceMinMaxKernel.cu:
```diff
@@ -178,9 +178,9 @@ REGISTER_DISPATCH(argmax_stub, &argmax_kernel_cuda);
 REGISTER_DISPATCH(min_stub, &min_kernel_impl);
 REGISTER_DISPATCH(max_stub, &max_kernel_impl);
-REGISTER_DISPATCH(_aminmax_stub, &_aminmax_kernel_impl);
+REGISTER_DISPATCH(minmax_stub, &minmax_kernel_impl);
```
This should be aminmax_stub and aminmax_kernel_impl?

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2701,16 +2701,29 @@
     CPU, CUDA: _aminmax
 
+- func: aminmax(Tensor self, *, int? dim=None, bool keepdim=False) -> (Tensor min, Tensor max)
+  device_check: NoCheck
```
Inline comment for why this check is skipped like here:

https://github.com/pytorch/pytorch/blob/725d98bab61d24509ffc199cee6d177f32a96e38/aten/src/ATen/native/native_functions.yaml#L208

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2701,16 +2701,29 @@
+  variants: function, method
+
+- func: aminmax.out(Tensor self, *, int? dim=None, bool keepdim=False, Tensor(a!) min, Tensor(b!) max) -> (Tensor(a!) min, Tensor(b!) max)
+  device_check: NoCheck
```
Inline comment here, too

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6361,6 +6361,63 @@ def merge_dicts(*dicts):
+add_docstr(torch.aminmax, r"""
+aminmax(input, *, dim=None, keepdim=False, out=None) -> (Tensor min, Tensor max)
+
+Computes the minimum and maximum values of the :attr:`input` tensor.
```
Great first sentence

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6361,6 +6361,63 @@ def merge_dicts(*dicts):
+    NaN values are propagated to the output if at least one value is NaN.
+
+.. seealso::
+    :func:`torch.amin` minimum value an input tensor
```
"computes just the minimum value"
"computes just the maximum value"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6361,6 +6361,63 @@ def merge_dicts(*dicts):
+    min=tensor(-3),
+    max=tensor(5))
+
+    >>> torch.aminmax(torch.tensor([1, -3, 5, torch.nan]))
```
Add a comment for this case like 

```
# aminmax propagates NaNs
```

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6361,6 +6361,63 @@ def merge_dicts(*dicts):
+    min=tensor(nan),
+    max=tensor(nan))
+
+    >>> x = torch.arange(10).view(2, 5)
```
Nit: prefer `t` when working with a single tensor in the docs and `a, b, c, ...` when working with multiple tensors that don't have natural names of their own

mruberry commented on torch/_torch_docs.py:
```diff
@@ -6361,6 +6361,63 @@ def merge_dicts(*dicts):
+    >>> x
+    tensor([[0, 1, 2, 3, 4],
+            [5, 6, 7, 8, 9]])
+    >>> x.aminmax(dim=0, keepdim=True)
```
Nice examples

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1865,12 +1865,29 @@ def sample_inputs_amax_amin(op_info, device, dtype, requires_grad, **kwargs):
-        ((), (0,), {'keepdim': True}),
+        ((), (), {'dim': 0, 'keepdim': True}),
+    )
+    return tuple(SampleInput((make_tensor(size, device, dtype,
```
Style nit: prefer starting with the for loop for readability (vs putting it at the end of the statement, which is yoda-like)

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6374,6 +6391,16 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
            supports_forward_ad=True,
            sample_inputs_func=sample_inputs_max_min_binary,),
+    OpInfo('aminmax',
+           ref=lambda x, dim=None, keepdim=False: (np.amin(x, axis=dim, keepdims=keepdim), np.amax(x, axis=dim, keepdims=keepdim)),
```
Cool bonus reference

mruberry commented:
Overall looks good with nicely updated tests, the port to structured, and well-written docs. Some minor comments inline. Build failures look real, however:

```
Jul 29 22:55:01 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/ReduceMinMaxKernel.cu(181): error: expected a type specifier
```

ngimel commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -118,6 +118,67 @@ TORCH_META_FUNC(argmin)
+  this->set_output(
+      0,
+      shape,
+      options.dtype(out0.defined() ? out0.scalar_type() : self.scalar_type()));
```
we don't support `out` of the different types (in fact, in the implementation there are checks to make sure that `min` and `max` are the same type. It means that you don't have to do `maybe_get_output`/`set_output`, just `set_output` is enough. 
cc @ezyang, we typically shouldn't need maybe_get_output in the META_FUNC, right?

heitorschueroff replied:
Wouldn't I still need to get one output to compute the dtype for set_output? Otherwise I will always set it to the input's dtype.

ezyang replied:
Yes, you only need it if you have behavior where you accommodate non-standard out dtype and handle the casting for them directly. If you require the out dtype to exactly match what the functional dtype was, only set_output is needed.

heitorschueroff committed:
Addressed PR comments on "Exposes _aminmax as aminmax and makes it structured"


This PR exposes the `torch._aminmax` operator as `torch.aminmax`.

**TODO**

- [x] add examples to documentation
- [x] add minmax to rst docs

fixes https://github.com/pytorch/pytorch/issues/62164

[ghstack-poisoned]

heitorschueroff committed:
Update on "Exposes _aminmax as aminmax and makes it structured"


This PR exposes the `torch._aminmax` operator as `torch.aminmax`.

**TODO**

- [x] add examples to documentation
- [x] add minmax to rst docs

fixes https://github.com/pytorch/pytorch/issues/62164

[ghstack-poisoned]

heitorschueroff committed:
rebase on "Exposes _aminmax as aminmax and makes it structured"


This PR exposes the `torch._aminmax` operator as `torch.aminmax`.

**TODO**

- [x] add examples to documentation
- [x] add minmax to rst docs

fixes https://github.com/pytorch/pytorch/issues/62164

[ghstack-poisoned]

heitorschueroff commented:
@mruberry @ngimel Thank you for the great comments. I updated the PR and it should be ready for another round of reviews.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2701,16 +2701,29 @@
 - func: _aminmax(Tensor self) -> (Tensor, Tensor)
   variants: function
   dispatch:
     CPU, CUDA: _aminmax_all
```
This is probably fine to leave these dispatches in but technically these are composite implicit autograd now, right?

heitorschueroff replied:
I will go through a deprecation cycle with these dispatches just to be safe in case someone discovered these and started using them. And yes, they are composite implicit autograd though aminmax does not yet support autograd.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1862,21 +1862,44 @@ def sample_inputs_take_along_dim(op_info, device, dtype, requires_grad, **kwargs
-        ((), (), {}),
-        ((), (0,), {'keepdim': True}),
+    # Ordered as (input shape, kwargs)
+    test_cases: Tuple[Tuple[int, ...], Dict[str, Any]] = (  # type: ignore[assignment]
```
I think the lint build is unhappy:

```
torch/testing/_internal/common_methods_invocations.py:1866:50: error: Name 'Any' is not defined  [name-defined]
```

I wouldn't bother typing your code here. If the lint build is finicky about typing for some reason then we should fix it to understand this file is untyped. But that's really for a separate PR so you can just tell lint to ignore lines it flags.

heitorschueroff replied:
I think it's just missing an import. VSCode has been imperfect at auto importing things for me haha.

mruberry reviewed (approved):
Awesome! Great quick turnaround on this popular request.

Just make sure to satisfy the lint build before merging this!

heitorschueroff committed:
Update on "Exposes _aminmax as aminmax and makes it structured"


This PR exposes the `torch._aminmax` operator as `torch.aminmax`.

**TODO**

- [x] add examples to documentation
- [x] add minmax to rst docs

fixes https://github.com/pytorch/pytorch/issues/62164

[ghstack-poisoned]

heitorschueroff committed:
Update on "Exposes _aminmax as aminmax and makes it structured"


This PR exposes the `torch._aminmax` operator as `torch.aminmax`.

**TODO**

- [x] add examples to documentation
- [x] add minmax to rst docs

fixes https://github.com/pytorch/pytorch/issues/62164

[ghstack-poisoned]

heitorschueroff committed:
Update on "Exposes _aminmax as aminmax and makes it structured"


This PR exposes the `torch._aminmax` operator as `torch.aminmax`.

**TODO**

- [x] add examples to documentation
- [x] add minmax to rst docs

fixes https://github.com/pytorch/pytorch/issues/62164

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30072246).

heitorschueroff committed:
Update on "Exposes _aminmax as aminmax and makes it structured"


This PR exposes the `torch._aminmax` operator as `torch.aminmax`.

**TODO**

- [x] add examples to documentation
- [x] add minmax to rst docs

fixes https://github.com/pytorch/pytorch/issues/62164

Differential Revision: [D30072246](https://our.internmc.facebook.com/intern/diff/D30072246)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Exposes _aminmax as aminmax and makes it structured"


This PR exposes the `torch._aminmax` operator as `torch.aminmax`.

**TODO**

- [x] add examples to documentation
- [x] add minmax to rst docs

fixes https://github.com/pytorch/pytorch/issues/62164

Differential Revision: [D30072246](https://our.internmc.facebook.com/intern/diff/D30072246)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30072246).

heitorschueroff committed:
Update on "Exposes _aminmax as aminmax and makes it structured"


This PR exposes the `torch._aminmax` operator as `torch.aminmax`.

**TODO**

- [x] add examples to documentation
- [x] add minmax to rst docs

fixes https://github.com/pytorch/pytorch/issues/62164

Differential Revision: [D30072246](https://our.internmc.facebook.com/intern/diff/D30072246)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30072246).

vadimkantorov commented:
Can it support multidim as well? Let's make a separate TODO issue for this?

heitorschueroff commented:
> Can it support multidim as well? Let's make a separate TODO issue for this?

It cannot support multidim yet. I am working on adding support for reduction kernels to enable this. There's already a general issue for this here https://github.com/pytorch/pytorch/issues/61582

ngimel commented:
It can support multidim on cuda (amin/amax do support multidim), but not on cpu.

heitorschueroff commented:
> It can support multidim on cuda (amin/amax do support multidim), but not on cpu.
> Edit: cuda kernel is written in a way that supports multidim, but current aminmax implementation doesn't accept multidim arg.

Thanks for pointing it out. I wanted to keep this PR focused on just exposing current implementation, which does not accept multi-dim. I am working on adding multi-dim support along with many other reductions that do not support it.
